src/Utility/TypeExtensions.cs

```diff
@@ -89,24 +90,34 @@ internal static object ToObject(this TypedData data)
                 case TypedData.DataOneofCase.None:
                     return null;
                 default:
-                    return new InvalidOperationException("Data Case was not set.");
+                    return new InvalidOperationException("DataCase was not set.");
             }
         }
 
-        private static JsonSerializerSettings setting =
-            new JsonSerializerSettings { TypeNameHandling = TypeNameHandling.None, MaxDepth = 3 };
-        private static object DeserializeJson(string json)
+        // PowerShell NuGet packages only have 'System.Management.Automation.dll' as the ref assembly, and thus types from other powershell assemblies
+        // cannot be used directly in an application that reference the PowerShell NuGet packages. This is tracked by PowerShell#8121.
+        // Here we need to use 'Microsoft.PowerShell.Commands.JsonObject' from 'Microsoft.PowerShell.Commands.Utility'. Due the above issue, we have to
+        // use reflection to call 'JsonObject.ConvertFromJson(...)'.
+        private const string UtilityAssemblyFullName = "Microsoft.PowerShell.Commands.Utility, Version=6.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35";
+        private static MethodInfo s_ConvertFromJson = null;
+        private static object ConvertFromJson(string json)
         {
-            var obj = JsonConvert.DeserializeObject(json, setting);
-            switch (obj)
+            if (s_ConvertFromJson == null)
             {
-                case JObject dict:
-                    return new Hashtable(dict.ToObject<Hashtable>(), StringComparer.OrdinalIgnoreCase);
-                case JArray list:
-                    return list.ToObject<object[]>();
-                default:
-                    return obj;
+                Assembly utilityAssembly = AppDomain.CurrentDomain.GetAssemblies().First(asm => asm.FullName == UtilityAssemblyFullName);
```

Comment:
If we put this in a static constructor, then when this type is accessed without having a powershell Runspace spined up, the reflection will fail because the `utility.dll` might not be loaded yet.